python/docs/tools.md

```diff
@@ -334,6 +336,87 @@ if __name__ == "__main__":
 
 _Source: [/python/examples/tools/wikipedia.py](/python/examples/tools/wikipedia.py)_
 
+### Using the `SandboxTool` (Python functions)
+
+If you want to use the Python function, use the [`SandboxTool`](/python/src/tools/code/sandbox.ts).
+
+<!-- embedme examples/tools/custom/sandbox.py -->
+
+```py
+import asyncio
+import os
+import sys
+import traceback
+
+from dotenv import load_dotenv
+
+from beeai_framework.errors import FrameworkError
+from beeai_framework.tools.code.sandbox import SandboxTool
+
+load_dotenv()
+
+
+async def main() -> None:
+    sandbox_tool = await SandboxTool.from_source_code(
+        url=os.getenv("CODE_INTERPRETER_URL", "http://127.0.0.1:50081"),
+        env={"API_URL": "https://riddles-api.vercel.app/random"},
+        source_code="""
+import requests
+import os
+from typing import Optional, Union, Dict
+
+def get_riddle() -> Optional[Dict[str, str]]:
+    '''
+    Fetches a random riddle from the Riddles API.
+
+    This function retrieves a random riddle and its answer. It does not accept any input parameters.
+
+    Returns:
+        Optional[Dict[str, str]]: A dictionary containing:
+            - 'riddle' (str): The riddle question.
+            - 'answer' (str): The answer to the riddle.
+        Returns None if the request fails.
+    '''
+    url = os.environ.get('API_URL')
+
+    try:
+        response = requests.get(url)
+        response.raise_for_status()
+        return response.json()
+    except Exception as e:
+        return None
+""",
+    )
+
+    result = await sandbox_tool.run({})
+
+    print(result.get_text_content())
+
+
+if __name__ == "__main__":
+    try:
+        asyncio.run(main())
+    except FrameworkError as e:
+        traceback.print_exc()
+        sys.exit(e.explain())
+
+```
+
+_Source: [examples/tools/custom/sandbox.py](/typescript/examples/tools/custom/sandbox.py)_
+
+> [!TIP]
+>
+> Environmental variables can be overridden (or defined) in the following ways:
+>
+> 1. During the creation of a `SandboxTool`, either via the constructor or the factory function (`SandboxTool.from_source_code`).
+> 2. By passing them directly as part of the options when invoking: `my_tool.run(..., env={ "MY_ENV": "MY_VALUE" })`.
+<!-- 3. Dynamically during execution via [`Emitter`](/python/docs/emitter.md): `my_tool.emitter.on("start", lambda data, event: (data.options.env.update({"MY_ENV": "MY_VALUE"})))`. -->
```

Comment:
I've left this commented out because currently this feature is actually broken, or at least mostly useless.

In our current implementation if the there are no options passed to Tool.run then it is set to None, whereas in TS it is set to empty dict. This means if there was never options initialized then there are no options to edit in the event data object.

Due to options being a generic class though this is a non-trivial fix to address this.

Reply:
Correct. Unfortunately, there is no easy fix for this.